benchmarks/overrides_benchmark/README.md

```diff
@@ -0,0 +1,39 @@
+# `__torch_function__` micro-benchmarks
+
+This benchmark suite provides a systemic way to measure the performance of `__torch_function__` overhead.
+
+## Getting started
+### Initial Setup
+Install `py-spy` by doing:
+
+```bash
+pip install py-spy
+```
+
+Note that more extensive documentation on using `py-spy` is available in `CONTRIBUTING.md`.
+
+### Running the benchmark
+Run one of the following commands in the terminal, with the working directory being `${PYTORCH_CLONE_DIR}/benchmarks/overrides_benchmark`:
+
+```bash
+# Benchmark all the cases
+python bench.py
+
+# Flame graph pertaining to each case.
+py-spy record -o tensor.svg --native -- python pyspybench.py Tensor
+py-spy record -o subtensor.svg --native -- python pyspybench.py SubTensor
+py-spy record -o overridden.svg --native -- python pyspybench.py WithTorchFunction
+```
+
+Here is a brief overview of what the results should look like, if run correctly:
+
+* Overhead for `torch` functions when run on `torch.Tensor` objects is close to 2 μs.
+* `__torch_function__` should add zero overhead for `torch.Tensor` inputs, a small overhead for subclasses of `torch.Tensor`, and a couple of microseconds for `Tensor`-likes with `__torch_function__`.
+* Changing the dispatching mechanism may result in changes that are on the order of 100 ns, which are hard to detect due to noise, but important.
+
+## Reporting benchmark results
+When modifying any of the machinery around `__torch_function__`, run the benchmark for both the feature branch and the point it diverges from `master`. For each of these:
+
+* Run `bench.py`, and include the output in your result.
+* For each case where `bench.py` shows a regression, run the commands described above, prefixing the output SVG filename (the input to the `-o` switch) with `base-` or `branch-` depending on the commit you are running the benchmark on.
+* For each SVG, open it in the browser, take a screenshot and include it in your result. Also include a ZIP file with all SVGs thus produced included.
```

Comment:
Couldn't this part be automated? Maybe rather than relying on the `py-spy` flame graph svg output it would make more sense to process the data from the "raw" output format to extract relevant information that can be copy/pasted into a PR discussion.

Reply:
Are there docs on that format somewhere?

Reply:
It’s mentioned in the output of `py-spy record —help`

Reply:
I mean the actual format that `raw` produces, is there a parser or some docs on it?

Reply:
No I don't think it's documented, py-spy's docs are pretty barebones. Looking at the py-spy source code it looks like it's writing the contents of the `HashMap` mapping stack frame names to counts to the output file, see:

https://github.com/benfred/py-spy/blob/d870b3768353001b31ee78129a79be99319b9bec/src/main.rs#L135-L145

https://github.com/benfred/py-spy/blob/d870b3768353001b31ee78129a79be99319b9bec/src/flamegraph.rs#L39-L42

Reply:
The py-spy raw output is in the 'folded stack' format used by flamegraph.pl (https://github.com/brendangregg/FlameGraph#2-fold-stacks) . The format is a text file with a stack trace /sample count per line, with each frame in the stack trace being delimited with a ';'.   Fwiw for benchmarking, you can take a 'before' and 'after' version of these raw files and compute a [differential flamegraph](http://www.brendangregg.com/blog/2014-11-09/differential-flame-graphs.html).

Reply:
I looked into this. It would introduce another dependency to rasterize the SVG, and another to diff the flame graph. I wonder if it's worth it.

Reply:
Not to mention auto checkout and auto builds -- Which is bad.